<a href="https://colab.research.google.com/github/Elizaluckianchikova/Homework-on-leetcode/blob/main/Lecture_5_(2023_10_18).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lecture 5: Data Structures (build-in)

> List

> Tuple

> Set

> Dictionary

In [3]:
# 1816. Truncate Sentence

# input
s = "Hello how are you Contestant"
k = 4

# solution
print(s.split())                   # string -> list
print(s.split()[0:k])              # slicing
print(" ".join(s.split()[0:k]))    # list -> string

['Hello', 'how', 'are', 'you', 'Contestant']
['Hello', 'how', 'are', 'you']
Hello how are you


### List

In [4]:
# index      0        1      2      3
my_list = ['Hello', 'how', 'are', 'you']

print(my_list)
print(my_list[0])
print(my_list[1])

print(my_list[1:3])

['Hello', 'how', 'are', 'you']
Hello
how
['how', 'are']


In [5]:
my_list.append("?")
print(my_list)

['Hello', 'how', 'are', 'you', '?']


In [6]:
my_list[0] = "Hi"
print(my_list)

['Hi', 'how', 'are', 'you', '?']


In [7]:
my_list_empty = []
print(my_list_empty)

[]


In [8]:
del my_list[0]
print(my_list)

['how', 'are', 'you', '?']


In [9]:
print(len(my_list))
print(type(my_list))

4
<class 'list'>


In [10]:
for i in range(len(my_list)):
    print(my_list[i])

how
are
you
?


In [11]:
for element in my_list:
    print(element)

how
are
you
?


In [12]:
# index      0        1      2      3
# index     -4       -3     -2     -1
my_list = ['Hello', 'how', 'are', 'you']

print(my_list[-4])
print(my_list[-4:-2])

Hello
['Hello', 'how']


In [13]:
# input
my_list = ['Hello', 'how', 'are', 'you', "?"]
my_list = ['Hello', 'how', 'are', 'you']

# solution
for i in range(len(my_list)):
    print(my_list[i])

Hello
how
are
you


In [14]:
print(list(range(4)))
len(my_list)

[0, 1, 2, 3]


4

In [15]:
print(list(range(10)))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]


In [16]:
range(len(my_list))

range(0, 4)

In [17]:
# list - изменяемая структура данных
my_list = [1, 2, 3]
print(my_list)
print(id(my_list))

print()
my_list.append(4)
print(my_list)
print(id(my_list))

[1, 2, 3]
134348767852800

[1, 2, 3, 4]
134348767852800


In [18]:
my_int = 3
print(my_int)
print(id(my_int))

print()
my_int += 12
print(my_int)
print(id(my_int))

3
134349196214576

15
134349196214960


In [19]:
my_list = [1, 2, 3]
print(my_list)

my_list_1 = my_list
print(my_list_1)

print()
my_list_1.append(4)
print(my_list_1)
print(my_list)

[1, 2, 3]
[1, 2, 3]

[1, 2, 3, 4]
[1, 2, 3, 4]


In [20]:
my_list = [1, 2, 3]
print(my_list)

my_list_1 = my_list.copy()
print(my_list_1)

print()
my_list_1.append(4)
print(my_list_1)
print(my_list)

[1, 2, 3]
[1, 2, 3]

[1, 2, 3, 4]
[1, 2, 3]


### Tuple

In [21]:
weekday_list = ['Понедельника', 'Вторника', 'Среды', 'Четверга', 'Пятницы', 'Субботы', 'Воскресенья']
weekday_tuple = ('Понедельника', 'Вторника', 'Среды', 'Четверга', 'Пятницы', 'Субботы', 'Воскресенья')

print(weekday_tuple[0])
print(weekday_tuple[0:3])

print()
for weekday in weekday_tuple:
    print(weekday)

Понедельника
('Понедельника', 'Вторника', 'Среды')

Понедельника
Вторника
Среды
Четверга
Пятницы
Субботы
Воскресенья


In [22]:
my_tuple = ()
print(my_tuple)

()


In [23]:
print(weekday_list.__sizeof__())
print(weekday_tuple.__sizeof__())

104
80


## Set

In [24]:
my_list = ["Mask", "Mask"]
print(my_list)

my_tuple = ("Mask", "Mask")
print(my_tuple)

['Mask', 'Mask']
('Mask', 'Mask')


In [25]:
my_set = {1, 2, 3}
print(my_set)

my_set.add(4)
print(my_set)

my_set.add(2)
print(my_set)

{1, 2, 3}
{1, 2, 3, 4}
{1, 2, 3, 4}


In [27]:
print(my_set[0])

TypeError: ignored

In [28]:
for element in my_set:
    print(element)

1
2
3
4


In [29]:
# set - изменяемая структура данных
my_set = {1, 2, 3}
print(my_set)
print(id(my_set))

print()
my_set.add(4)
print(my_set)
print(id(my_set))

{1, 2, 3}
134348769943936

{1, 2, 3, 4}
134348769943936


### Dictionary

In [30]:
# index        -2            -1
# index         0             1
my_list = [[5, 10, 15], [20, 30, 40]]
print(my_list)
print(my_list[0])

[[5, 10, 15], [20, 30, 40]]
[5, 10, 15]


In [31]:
my_list = [[5, 10, 15], [20, 30, 40]]
print(my_list[0])

# пары key:value
my_dict = {0: [5, 10, 15], 1: [20, 30, 40]}
print(my_dict[0])

[5, 10, 15]
[5, 10, 15]


In [32]:
my_dict = {
    "MIPT": ["Биология", "Прикладная математика", "Информатика"],
    "СПбГУ": ["Биология", "Прикладная математика", "Информатика"]
}

print(my_dict["MIPT"])

['Биология', 'Прикладная математика', 'Информатика']


In [33]:
print(my_dict)
print()

my_dict["MIPT_2"] = ["Вычислительная биология"]
print(my_dict)
print()

my_dict["MIPT"] = ["Биология"]
print(my_dict)

{'MIPT': ['Биология', 'Прикладная математика', 'Информатика'], 'СПбГУ': ['Биология', 'Прикладная математика', 'Информатика']}

{'MIPT': ['Биология', 'Прикладная математика', 'Информатика'], 'СПбГУ': ['Биология', 'Прикладная математика', 'Информатика'], 'MIPT_2': ['Вычислительная биология']}

{'MIPT': ['Биология'], 'СПбГУ': ['Биология', 'Прикладная математика', 'Информатика'], 'MIPT_2': ['Вычислительная биология']}


In [34]:
del my_dict["MIPT_2"]
print(my_dict)

{'MIPT': ['Биология'], 'СПбГУ': ['Биология', 'Прикладная математика', 'Информатика']}


In [35]:
del my_dict["MIPT_2"]
print(my_dict)

KeyError: ignored

# Homework

In [36]:
# 1. Richest Customer Wealth (LeetCode: 1672)
# https://leetcode.com/problems/richest-customer-wealth/description/

# 2. Number of Good Pairs (LeetCode: 1512)
# https://leetcode.com/problems/number-of-good-pairs/description/

# 3. Check If Two String Arrays are Equivalent (LeetCode: 1662)
# https://leetcode.com/problems/check-if-two-string-arrays-are-equivalent/description/

# 4. Pascal's Triangle (LeetCode: 118)
# https://leetcode.com/problems/pascals-triangle/

# 5. Pascal's Triangle II (LeetCode: 119)
# https://leetcode.com/problems/pascals-triangle-ii/

# 6. Valid Anagram (LeetCode: 242)
# https://leetcode.com/problems/valid-anagram/description/

# 7. Find All Duplicates in an Array (LeetCode: 442)
# https://leetcode.com/problems/find-all-duplicates-in-an-array/description/

# 8. Check if All Characters Have Equal Number of Occurrences (LeetCode: 1941)
# https://leetcode.com/problems/check-if-all-characters-have-equal-number-of-occurrences/

# 9. Check if the Sentence Is Pangram (LeetCode: 1832)
# https://leetcode.com/problems/check-if-the-sentence-is-pangram/

# 10. Two Sum (LeetCode: 1)
# https://leetcode.com/problems/two-sum/

**1672. Richest Customer Wealth**
You are given an m x n integer grid accounts where accounts[i][j] is the amount of money the i​​​​​​​​​​​th​​​​ customer has in the j​​​​​​​​​​​th​​​​ bank. Return the wealth that the richest customer has.

A customer's wealth is the amount of money they have in all their bank accounts. The richest customer is the customer that has the maximum wealth.



Example 1:

Input: accounts = [[1,2,3],[3,2,1]]
Output: 6
Explanation:
1st customer has wealth = 1 + 2 + 3 = 6
2nd customer has wealth = 3 + 2 + 1 = 6
Both customers are considered the richest with a wealth of 6 each, so return 6.
Example 2:

Input: accounts = [[1,5],[7,3],[3,5]]
Output: 10
Explanation:
1st customer has wealth = 6
2nd customer has wealth = 10
3rd customer has wealth = 8
The 2nd customer is the richest with a wealth of 10.
Example 3:

Input: accounts = [[2,8,7],[7,1,3],[1,9,5]]
Output: 17


Constraints:

m == accounts.length
n == accounts[i].length
1 <= m, n <= 50
1 <= accounts[i][j] <= 100

In [ ]:
class Solution(object):
    def maximumWealth(self, accounts):
        """
        :type accounts: List[List[int]]
        :rtype: int
        """
        return max(sum(account) for account in accounts)

**1512. Number of Good Pairs**

Given an array of integers nums, return the number of good pairs.

A pair (i, j) is called good if nums[i] == nums[j] and i < j.



Example 1:

Input: nums = [1,2,3,1,1,3]
Output: 4
Explanation: There are 4 good pairs (0,3), (0,4), (3,4), (2,5) 0-indexed.
Example 2:

Input: nums = [1,1,1,1]
Output: 6
Explanation: Each pair in the array are good.
Example 3:

Input: nums = [1,2,3]
Output: 0


Constraints:

1 <= nums.length <= 100
1 <= nums[i] <= 100

In [36]:
class Solution(object):
    def numIdenticalPairs(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        ans = 0
        cnt = Counter()
        for x in nums:
            ans += cnt[x]
            cnt[x] += 1
        return ans

**1662. Check If Two String Arrays are Equivalent**

Given two string arrays word1 and word2, return true if the two arrays represent the same string, and false otherwise.

A string is represented by an array if the array elements concatenated in order forms the string.



Example 1:

Input: word1 = ["ab", "c"], word2 = ["a", "bc"]
Output: true
Explanation:
word1 represents string "ab" + "c" -> "abc"
word2 represents string "a" + "bc" -> "abc"
The strings are the same, so return true.
Example 2:

Input: word1 = ["a", "cb"], word2 = ["ab", "c"]
Output: false
Example 3:

Input: word1  = ["abc", "d", "defg"], word2 = ["abcddefg"]
Output: true


Constraints:

1 <= word1.length, word2.length <= 103
1 <= word1[i].length, word2[i].length <= 103
1 <= sum(word1[i].length), sum(word2[i].length) <= 103
word1[i] and word2[i] consist of lowercase letters.

In [ ]:
#solution_1
class Solution(object):
    def arrayStringsAreEqual(self, word1, word2):
        """
        :type word1: List[str]
        :type word2: List[str]
        :rtype: bool
        """
        return ''.join(word1) == ''.join(word2)

In [38]:
#solution_2
class Solution(object):
    def arrayStringsAreEqual(self, word1, word2):
        """
        :type word1: List[str]
        :type word2: List[str]
        :rtype: bool
        """
        i = j = x = y = 0
        while i < len(word1) and j < len(word2):
            if word1[i][x] != word2[j][y]:
                return False
            x, y = x + 1, y + 1
            if x == len(word1[i]):
                x, i = 0, i + 1
            if y == len(word2[j]):
                y, j = 0, j + 1
        return i == len(word1) and j == len(word2)

**118. Pascal's Triangle**

Given an integer numRows, return the first numRows of Pascal's triangle.

In Pascal's triangle, each number is the sum of the two numbers directly above it as shown:




Example 1:

Input: numRows = 5
Output: [[1],[1,1],[1,2,1],[1,3,3,1],[1,4,6,4,1]]
Example 2:

Input: numRows = 1
Output: [[1]]


Constraints:

1 <= numRows <= 30

In [ ]:
class Solution:
    def generate(self, numRows):
        triangles = []
        for i in range(numRows):
            triangles.append([])
            for j in range(i + 1):
                if j == 0 or j == i:
                    triangles[i].append(1)
                else:
                    triangles[i].append(triangles[i - 1][j - 1] + triangles[i - 1][j])
        return triangles

**242. Valid Anagram**

Given two strings s and t, return true if t is an anagram of s, and false otherwise.

An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.



Example 1:

Input: s = "anagram", t = "nagaram"
Output: true
Example 2:

Input: s = "rat", t = "car"
Output: false


Constraints:

1 <= s.length, t.length <= 5 * 104
s and t consist of lowercase English letters.

In [ ]:
class Solution(object):
    def isAnagram(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        if len(s) != len(t):
            return False

        for l in set(s): # loop count <= 26
            if s.count(l) != t.count(l):
                return False

        return True


**442. Find All Duplicates in an Array**

Given an integer array nums of length n where all the integers of nums are in the range [1, n] and each integer appears once or twice, return an array of all the integers that appears twice.

You must write an algorithm that runs in O(n) time and uses only constant extra space.



Example 1:

Input: nums = [4,3,2,7,8,2,3,1]
Output: [2,3]
Example 2:

Input: nums = [1,1,2]
Output: [1]
Example 3:

Input: nums = [1]
Output: []


Constraints:

n == nums.length
1 <= n <= 105
1 <= nums[i] <= n
Each element in nums appears once or twice.

In [ ]:
class Solution:
    def findDuplicates(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        # 1 2 3 4 5 6
        # 0 1 2 3 4 5
        ans = []
        for n in nums:
            nums[abs(n) - 1] = -nums[abs(n) - 1]
            if nums[abs(n) - 1] > 0:
                ans.append(abs(n))
        return ans

**1941. Check if All Characters Have Equal Number of Occurrences**

Given a string s, return true if s is a good string, or false otherwise.

A string s is good if all the characters that appear in s have the same number of occurrences (i.e., the same frequency).



Example 1:

Input: s = "abacbc"
Output: true
Explanation: The characters that appear in s are 'a', 'b', and 'c'. All characters occur 2 times in s.
Example 2:

Input: s = "aaabb"
Output: false
Explanation: The characters that appear in s are 'a' and 'b'.
'a' occurs 3 times while 'b' occurs 2 times, which is not the same number of times.


Constraints:

1 <= s.length <= 1000
s consists of lowercase English letters.

In [ ]:
class Solution(object):
    def areOccurrencesEqual(self, s):
        """
        :type s: str
        :rtype: bool
        """
        cnt = Counter(s)
        return len(set(cnt.values())) == 1


**1832. Check if the Sentence Is Pangram**

A pangram is a sentence where every letter of the English alphabet appears at least once.

Given a string sentence containing only lowercase English letters, return true if sentence is a pangram, or false otherwise.



Example 1:

Input: sentence = "thequickbrownfoxjumpsoverthelazydog"
Output: true
Explanation: sentence contains at least one of every letter of the English alphabet.
Example 2:

Input: sentence = "leetcode"
Output: false


Constraints:

1 <= sentence.length <= 1000
sentence consists of lowercase English letters.

In [ ]:
class Solution(object):
    def checkIfPangram(self, sentence):
        """
        :type sentence: str
        :rtype: bool
        """
        return len(set(sentence)) == 26

**1. Two Sum**

Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.



Example 1:

Input: nums = [2,7,11,15], target = 9
Output: [0,1]
Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
Example 2:

Input: nums = [3,2,4], target = 6
Output: [1,2]
Example 3:

Input: nums = [3,3], target = 6
Output: [0,1]


Constraints:

2 <= nums.length <= 104
-109 <= nums[i] <= 109
-109 <= target <= 109
Only one valid answer exists.

In [ ]:
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        seen = {}
        for i, num in enumerate(nums):

            complement = target - num
            if complement in seen:
                return [seen[complement], i]
            seen[num] = i
        return None

**167. Two Sum II - Input Array Is Sorted**

Given a 1-indexed array of integers numbers that is already sorted in non-decreasing order, find two numbers such that they add up to a specific target number. Let these two numbers be numbers[index1] and numbers[index2] where 1 <= index1 < index2 < numbers.length.

Return the indices of the two numbers, index1 and index2, added by one as an integer array [index1, index2] of length 2.

The tests are generated such that there is exactly one solution. You may not use the same element twice.

Your solution must use only constant extra space.



Example 1:

Input: numbers = [2,7,11,15], target = 9
Output: [1,2]
Explanation: The sum of 2 and 7 is 9. Therefore, index1 = 1, index2 = 2. We return [1, 2].
Example 2:

Input: numbers = [2,3,4], target = 6
Output: [1,3]
Explanation: The sum of 2 and 4 is 6. Therefore index1 = 1, index2 = 3. We return [1, 3].
Example 3:

Input: numbers = [-1,0], target = -1
Output: [1,2]
Explanation: The sum of -1 and 0 is -1. Therefore index1 = 1, index2 = 2. We return [1, 2].


Constraints:

2 <= numbers.length <= 3 * 104
-1000 <= numbers[i] <= 1000
numbers is sorted in non-decreasing order.
-1000 <= target <= 1000
The tests are generated such that there is exactly one solution.

In [ ]:
class Solution(object):
    def twoSum(self, numbers, target):
        """
        :type numbers: List[int]
        :type target: int
        :rtype: List[int]
        """
        i = 0
        j = len(numbers) - 1

        while i < j:
            if numbers[i] + numbers[j] < target:
                i += 1
            elif numbers[i] + numbers[j] > target:
                j -= 1
            else:
                return [i + 1, j + 1]


        return []